Allow different parameters from config files

K-Meech opened the pull request:
For https://github.com/HEFTIEProject/zarr-benchmarks/issues/33

This PR is progress towards creating the [three types of plots we agreed on](https://github.com/HEFTIEProject/zarr-benchmarks/issues/33#issuecomment-2796914960). Previously, we read parameter combinations from a python file `tests/benchmarks/benchmark_parameters.py`, and benchmarks would run for all combinations of these parameters. The problem is there are some situations where we don't want to run for all combinations e.g. the 'chunk_size' plot where we need all chunk sizes from 60-130, with only **one** compressor / compression level.

We could expand the benchmark code to hardcode these situations, but it seemed easier to create config files for each main type of plot. These are under `tests/benchmarks/benchmark_configs` as json files, along with a json schema at `tests/benchmark_configs/schema` allowing automatic validation of e.g. provided compression level ranges, blosc compressor names etc...

The config name is provided on the commandline e.g.
```
tox -- --dev-image --rounds=1 --warmup-rounds=0 --config=shuffle
```

The downside is we would need to run tox three times to create all the combinations we need for plots:
```
tox -- --dev-image --rounds=1 --warmup-rounds=0 --config=shuffle
tox -- --dev-image --rounds=1 --warmup-rounds=0 --config=chunk_size
tox -- --dev-image --rounds=1 --warmup-rounds=0 --config=read_write
```
We could add these three commands to the tox config for each environment, but there didn't seem to be an easy way to override this to just run one configuration (this would be useful for dev work, as even with `--dev-image`, all of these combinations take a while to run!). It might be simpler to provide some kind of bash script (or similar) to run all the required tox commands and save the output to appropriate locations.

Any suggestions / input is very welcome! (opening as draft as I still need to add info to the readme, but wanted some feedback first)

K-Meech commented:
Thanks @dstansby. I've heard of `pydantic`, but never used it before - good to know that it can generate json schemas! I'll make a note on an issue so we can swap to pydantic later, if we have time :smile:

I like the `--config=all` idea, but I'm not sure how would be best to implement this? All the additional arguments to tox (like `--config`), get passed directly to the underlying `pytest` commands and eventually to `pytest_generate_tests`. I could process `all` at this stage, and turn it into a list of config files in the directory, but I can only add one set of parameters to each benchmark with `metafunc.parametrize`. I somehow need to run `pytest` multiple times, once for each config file, but I'm not sure `tox` supports this kind of conditional logic based on command-line arguments?

K-Meech commented on tests/conftest.py:
```diff
@@ -65,3 +73,27 @@ def image(dev_image):
+    config = read_json_file(config_path)
+
+    for key, values in config.items():
+        if key not in metafunc.fixturenames:
```
I think this `continue` is necessary to ensure that paramaterization isn't added to tests that don't have the relevant key in their arguments. e.g. under `tests/tests/test_write_zarr_array` I have some tests that check the `write_empty_chunks` behaviour. Without this continue, they would have paramaterization added for `chunk_size`, `blosc_clevel` etc....

dstansby commented:
Ah right, I see what you mean. Maybe best to leave it like this in this PR, and then try and find a solution later if it's annoying having to run them separately.

K-Meech commented:
I've been playing around with a few options this morning, and I think I have one that works now. Instead of adding parameters inside `pytest_generate_tests` like:
```
@pytest.mark.parametrize("chunk_size", [values..])
@pytest.mark.parametrize("blosc_clevel", [values..)
def test...
```
which locks us into allxall combinations. I'm now adding them like:
```
@pytest.mark.parametrize("chunk_size, blosc_clevel", [values..])
def test...
```
This allows me to parse all the config files, generate the combinations required for each and add them all as parameters in one go. This means all the results will be combined into one big json file - but it seems to be working quite nicely at the moment.

What do you think? Happy to remove this if you think it's overkill (everything I added is in the last commit on this branch).

dstansby commented on tests/conftest.py:
```diff
@@ -65,3 +74,60 @@ def image(dev_image):
+        parametrize_values.update(parameter_combinations)
+
+    # sort values for parametrize, so they are more readable in pytest output
+    parametrize_values = list(parametrize_values)
+    parametrize_values.sort()
```
```suggestion
    parametrize_values = sorted(parametrize_values)
```

dstansby commented on tests/tests/test_configs.py:
```diff
@@ -0,0 +1,19 @@
+def test_schemas_valid():
+    """Validate all jsons in the 'benchmarks_configs' dir against the schema"""
+
+    configs_dir = Path("tests/benchmarks/benchmark_configs")
```
```suggestion
    configs_dir = Path(__file__).parent.parent / "benchmarks" / "benchmark_configs"
```
Probably best to use a relative path to the directory, just in case this file is imported from a different working directory.

dstansby commented on tests/tests/test_configs.py:
```diff
@@ -0,0 +1,19 @@
+    configs_dir = Path("tests/benchmarks/benchmark_configs")
+    schema = utils.read_json_file(configs_dir / "schema" / "benchmark.schema.json")
+
+    for config_file in configs_dir.iterdir():
+        if not config_file.is_file():
+            continue
```
```suggestion
    for config_file in configs_dir.glob("*.json"):
```

dstansby commented on tests/conftest.py:
```diff
@@ -1,12 +1,21 @@
+    parser.addoption(
+        "--config",
+        action="store",
+        default="dev",
+        type=str,
+        help="name of config json file from tests/benchmarks/benchmark_configs",
```
```suggestion
        default="dev",
        type=str,
        help="Name of config json file from tests/benchmarks/benchmark_configs, or one of {'all', 'dev'}",
```

dstansby commented on tests/conftest.py:
```diff
@@ -65,3 +74,60 @@ def image(dev_image):
+    """Parse json config files. 'all' will parse every config in tests/benchmark_configs (except for dev, which contains
+    a small set of parameters for development runs)."""
+
+    configs_dir = pathlib.Path("tests/benchmarks/benchmark_configs")
```
```suggestion
    configs_dir = Path(__file__).parent / "benchmarks" / "benchmark_configs"
```

dstansby commented on tests/conftest.py:
```diff
@@ -65,3 +74,60 @@ def image(dev_image):
+    configs = []
+
+    if config_name == "all":
+        for config_file in configs_dir.iterdir():
```
If instead you want a one-liner:
```
configs = [expand_min_max(read_json_file(config_file)) for config file in [f for f in configs_dir.glob("*.json") if f.stem != 'dev']]
```

but now I've written it it seems a bit silly 😄

dstansby commented on tests/conftest.py:
```diff
@@ -65,3 +74,60 @@ def image(dev_image):
+    configs = []
+
+    if config_name == "all":
+        for config_file in configs_dir.iterdir():
+            if config_file.is_file() and config_file.stem != "dev":
```
```suggestion
        for config_file in configs_dir.glob("*.json"):
            if config_file.stem != "dev":
```
here's a slightly saner option that still simplifies the code a bit

dstansby commented on tests/conftest.py:
```diff
@@ -65,3 +74,60 @@ def image(dev_image):
+            if config_file.is_file() and config_file.stem != "dev":
+                config = expand_min_max(read_json_file(config_file))
+                configs.append(config)
+    else:
```
Should there also be a branch for 'dev' here? Or am I misunderstanding?

K-Meech replied:
Thanks for the input @dstansby ! 

I don't think it's necessary to have another branch - `dev` is a config file like any of the others in `tests/benchmarks/benchmark_configs`, it just contains a small number of options for each parameter so it runs quickly (it's also the default option if no `config` is provided).

So running e.g. the `shuffle` config:
```
tox -- --dev-image --config=shuffle
```

Works the same as running the `dev` config:
```
tox -- --dev-image --config=dev
```

I've excluded `dev` from `--all`, as it doesn't contain parameter combinations we'll actually use in plots.

dstansby replied:
That makes sense 👍

dstansby commented on tests/conftest.py:
```diff
@@ -65,3 +74,60 @@ def image(dev_image):
+    return configs
+
+
+def pytest_generate_tests(metafunc):
```
Could you type this? It would help me (at least) understand what's going on.

K-Meech replied:
Will do! I'm mainly following the [`pytest_generate_tests` example from the parametrization docs](https://docs.pytest.org/en/stable/how-to/parametrize.html#basic-pytest-generate-tests-example)

dstansby commented on tests/conftest.py:
```diff
@@ -65,3 +74,60 @@ def image(dev_image):
+    configs = parse_config_files(config_name)
+
+    # keys from the config, that are used as arguments for this function
+    used_config_keys = [key for key in configs[0] if key in metafunc.fixturenames]
+    if len(used_config_keys) == 0:
+        return
```
In what case would there be keys in the config that aren't used in the test functions?

K-Meech replied:
`pytest_generate_tests` is called in `pytest`'s collection phase, once for every test function it collects.

If we run the benchmarks like below (i.e. without `benchmark-only`):
```
tox -- --dev-image --rounds=1 --warmup-rounds=0
```
It will run the tests under `tests/tests`, as well as the benchmarks under `tests/benchmarks`. When a test is collected e.g. those under `tests/tests/test_write_zarr_array`, they will pass through `pytest_generate_tests`. As they use none of the benchmark parameters, they will trigger the `if len(used_config_keys) == 0` to skip parametrization.

dstansby reviewed (commented):
New code looks 👍 , I like the approach. I left a couple of questions, and some suggestions to simplify code.

K-Meech commented:
I think I've addressed all of your comments now @dstansby + I updated the readme to include the `--config` options. I re-arranged a few sections in the readme under sub-headings, as it was getting rather long for one section.

dstansby commented on README.md:
````diff
@@ -27,47 +29,71 @@ pip install .
+# zarr-python v3 only
+tox run -e py313-zarrv3
+```
+
````
```suggestion

To see a list of available environments, use `tox -l`.
```

dstansby reviewed (approved):
Wonderful, this looks great to me - just one minor suggetion to add to the README, feel free to take or leave and self merge! 🚢
